crates/ibc/src/clients/ics07_tendermint/client_state.rs

```diff
@@ -342,262 +263,111 @@ impl Ics2ClientState for ClientState {
-                    _ => ClientError::Other {
-                        description: e.to_string(),
-                    },
-                })?
-                .as_ref()
-                .map(|cs| downcast_tm_consensus_state(cs.as_ref()))
-                .transpose()?;
-
-            if let Some(next_cs) = maybe_next_cs {
-                // New (untrusted) header timestamp cannot occur after next
-                // consensus state's height
-                if header.signed_header.header().time > next_cs.timestamp {
-                    return Err(ClientError::ClientSpecific {
-                        description: Error::HeaderTimestampTooHigh {
-                            actual: header.signed_header.header().time.to_string(),
-                            max: next_cs.timestamp.to_string(),
-                        }
-                        .to_string(),
-                    });
-                }
-            }
-        }
+        let updated_heights = vec![header_height];
+        Ok(updated_heights)
+    }
 
-        // (cs-trusted, cs-prev, cs-new)
-        if header.trusted_height < header.height() {
-            let maybe_prev_cs = ctx
-                .prev_consensus_state(&client_id, &header.height())
-                .map_err(|e| match e {
-                    ContextError::ClientError(e) => e,
-                    _ => ClientError::Other {
-                        description: e.to_string(),
-                    },
-                })?
-                .as_ref()
-                .map(|cs| downcast_tm_consensus_state(cs.as_ref()))
-                .transpose()?;
-
-            if let Some(prev_cs) = maybe_prev_cs {
-                // New (untrusted) header timestamp cannot occur before the
-                // previous consensus state's height
-                if header.signed_header.header().time < prev_cs.timestamp {
-                    return Err(ClientError::ClientSpecific {
-                        description: Error::HeaderTimestampTooLow {
-                            actual: header.signed_header.header().time.to_string(),
-                            min: prev_cs.timestamp.to_string(),
-                        }
-                        .to_string(),
-                    });
-                }
-            }
-        }
+    fn update_state_on_misbehaviour(
+        &self,
+        ctx: &mut dyn ExecutionContext,
+        client_id: &ClientId,
+        _client_message: Any,
+        _update_kind: &UpdateClientKind,
+    ) -> Result<(), ClientError> {
+        let frozen_client_state = self
+            .clone()
+            .with_frozen_height(Height::new(0, 1).unwrap())
+            .into_box();
 
-        Ok(UpdatedState {
-            client_state: client_state.with_header(header.clone())?.into_box(),
-            consensus_state: TmConsensusState::from(header).into_box(),
-        })
+        ctx.store_client_state(ClientStatePath::new(client_id), frozen_client_state)?;
```

Comment:
The justification for this is [here](https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-006-02-client-refactor.md#new-clients-may-want-to-do-batch-updates) in the ADR (and links to a [concrete reason](https://github.com/cosmos/ibc-go/issues/284#issuecomment-1005583679) why this is useful for polkadot).

> Anywhere else, when a new provable state is obtained upon a process, it is only set within an execute() handler. 

The model of "1 update client message - 1 client/consensus state update" no longer applies for `MsgUpdateClient` and `MsgSubmitMisbehaviour`. We want the light clients to perform the state updates to enable them to do it on their own schedule.

> This function can return new states (e.g. using UpdatedState),

What do you mean? The function no longer returns `UpdatedState`.

> and not leave the storing to users and hope they'll do it right.

In ibc-rs we have 2 types of "users": light client devs, and chain devs. Light clients are part of the ibc-rs system; if you use a buggy light client, then yes the whole system will be buggy. We happen to be the ones developing the tendermint light client, but the goal is to enable other teams to write light clients as well, and yes we rely on them to write quality light clients. Ultimately, it is always the chain dev's responsibility to audit any code they choose to integrate.

Note that this is not different from any other concern taken care of by light clients; we previously relied on the light client to give us the appropriate `UpdatedState`. Us storing it in `execute()` ourselves is a false sense of security; we still need the light client to give us the proper object to store. All that is changed is that the light client now controls not only the object that will be stored, but also *when* it will be stored.

Reply:
opened #615